src/facade/dragonfly_connection.cc

```diff
@@ -101,21 +103,74 @@ struct Connection::Shutdown {
   }
 };
 
+// Used as custom deleter for Request object
+struct Connection::RequestDeleter {
+  void operator()(Request* req) const;
+};
+
 struct Connection::Request {
-  absl::FixedArray<MutableSlice, 6> args;
+  struct PipelineMsg {
+    absl::FixedArray<MutableSlice, 6> args;
 
-  // I do not use mi_heap_t explicitly but mi_stl_allocator at the end does the same job
-  // of using the thread's heap.
-  // The capacity is chosen so that we allocate a fully utilized (256 bytes) block.
-  absl::FixedArray<char, kReqStorageSize, mi_stl_allocator<char>> storage;
-  AsyncMsg* async_msg = nullptr;  // allocated and released via mi_malloc.
+    // I do not use mi_heap_t explicitly but mi_stl_allocator at the end does the same job
+    // of using the thread's heap.
+    // The capacity is chosen so that we allocate a fully utilized (256 bytes) block.
+    absl::FixedArray<char, kReqStorageSize, mi_stl_allocator<char>> storage;
+
+    PipelineMsg(size_t nargs, size_t capacity) : args(nargs), storage(capacity) {
+    }
+  };
 
-  Request(size_t nargs, size_t capacity) : args(nargs), storage(capacity) {
+ private:
+  using MessagePayload = std::variant<PipelineMsg, AsyncMsg>;
+
+  Request(size_t nargs, size_t capacity) : payload(PipelineMsg{nargs, capacity}) {
+  }
+
+  Request(AsyncMsg msg) : payload(std::move(msg)) {
   }
 
   Request(const Request&) = delete;
+
+ public:
+  static RequestPtr New(mi_heap_t* heap, RespVec args, size_t capacity);
+
+  static RequestPtr New(const PubMessage& pub_msg, fibers_ext::BlockingCounter bc);
+
+  MessagePayload payload;
 };
 
+Connection::RequestPtr Connection::Request::New(mi_heap_t* heap, RespVec args, size_t capacity) {
+  constexpr auto kReqSz = sizeof(Request);
+  void* ptr = mi_heap_malloc_small(heap, kReqSz);
+  Request* msg = new (ptr) Request(args.size(), capacity);
```

Comment:
I would call it `req` here and `pipeline_msg` below - less confusing

Reply:
done